### PR TITLE
Nshm organisational refactor

### PR DESCRIPTION
**Pull request by lispandfound**
This PR requires https://github.com/ucgmsim/qcore/pull/305 and https://github.com/ucgmsim/qcore/pull/303.

This pull request represents my current vision for how I think the nshmdb should work going forward. The purpose of this tool is as follows:

1. Provide a script to produce an SQLite database to efficiently query fault geometry (and other things going forward).
2. To provide any code relying on NSHM2022 data a simple API to work with.
3. To provide missing visualisation tools for NSHM (essentially, plugging gaps in the rupture explorer).

To this end we now have three modules and one script.

- nshmdb. This module provides an interface to query the database.
- fault. Contains the classes defining a fault geometry.
- plotting.rupture. The plotting package contains extra visualisation tools, for now just a tool to visualise a rupture scenario with pygmt.

The script has also been improved. It now uses the Python `zipfile` module to read the CRU solution zip file directly (i.e. without the user specifying the paths of the individial files). It does not extract need an extracted zip either, it's all read in binary. This means that the user does not need to understand the CRU solution zip's many undocumented files to generate the database. It is as simple as downloading the CRU solution zip file and running:

```bash
$ python nshmdb/scripts/nshm_db_generator.py ~/Downloads/CRU_fault_system_solution.zip nshmdb.db
```
or if they've installed the package the script lives in their path so this also works
```bash
$ nshm-db-generator ~/Downloads/CRU_fault_system_solution.zip nshmdb.db
```

The script now gives live updates via tqdm on the database build progress (which varies from 5 minutes on my home desktop, to 15 minutes on my work desktop).

Here is an example of how you can use the new API:

```python
>>> from nshmdb import nshmdb
>>> from nshmdb.plotting import rupture
>>> db = nshmdb.NSHMDB('nshmdb.db')
>>> rupture_100 = db.get_rupture_faults(100)
>>> akatarawa = rupture_100[0]
>>> plane = akatarawa.planes[0]
>>> np.round(plane.dip)
75.0
>>> plane.bottom_m
20000.0
>>> rupture.plot_rupture('Rupture 100', rupture_100)
# see plot below
```

The above code produces the following plot of the faults involved in the rupture.

image

### CODE REVIEW

File: nshmdb/fault.py
Comment (joelridden) [inline nshmdb/fault.py]: from qcore import coordinates, geo

File: nshmdb/fault.py
Comment (joelridden) [inline nshmdb/fault.py]: projected_width_m?

File: nshmdb/nshmdb.py
Comment (joelridden) [inline nshmdb/nshmdb.py]: Probably should raise a file not found error?

File: nshmdb/plotting/rupture.py
Comment (joelridden) [inline nshmdb/plotting/rupture.py]: idk if fig.show() is great here? Maybe either return the figure or could save to a file?

File: nshmdb/scripts/nshm_db_generator.py
Comment (joelridden) [inline nshmdb/scripts/nshm_db_generator.py]: This isnt used?

File: pyproject.toml
Comment (joelridden) [inline pyproject.toml]: Where we gonna do a requirements.txt instead of this to allow for everyone to decide how to do their environments?

**Review by joelridden — COMMENTED** [review]
Decent work! Not that strong on some of the math side for some of the fault calculations but I'm sure it works. Just a few minor comments.

Love the documentation

**Review by AndrewRidden-Harper — DISMISSED** [review]

File: nshmdb/nshmdb.py
Comment (sungeunbae) [inline nshmdb/nshmdb.py:67]: Agreed

File: nshmdb/plotting/rupture.py
Comment (sungeunbae) [reply nshmdb/plotting/rupture.py]: Saving to a file sounds better

File: nshmdb/fault.py
Comment (claudio525) [inline nshmdb/fault.py:171]: Isn't this always strike + 90?

Comment (claudio525) [reply nshmdb/fault.py:171]: Also where do you use this? Just curious since I have never come across it being used

File: nshmdb/fault.py
Comment (claudio525) [inline nshmdb/fault.py:207]: Just to be consistent, I think we usually describe it as "along strike" and "down dip"

File: nshmdb/fault.py
Comment (lispandfound) [reply nshmdb/fault.py:171]: Solvis defines DipDir in the json file. Here is an example entry: 

```json
{
      "type": "Feature",
      "id": 0,
      "properties": {
        "FaultID": 0,
        "FaultName": "Acton, Subsection 0",
        "DipDeg": 60.0,
        "Rake": 110.0,
        "LowDepth": 27.12,
        "UpDepth": 0.0,
        "DipDir": 94.3,
        "AseismicSlipFactor": 0.0,
        "CouplingCoeff": 1.0,
        "SlipRate": 1.1875,
        "ParentID": 0,
        "ParentName": "Acton",
        "SlipRateStdDev": 0.50237
      },
      "geometry": {
        "type": "LineString",
        "coordinates": [
            ...
        ]
      }
    }
``` 

It's because the fault segments are not simple rectangles anymore, but instead parallelograms.

File: nshmdb/fault.py
Comment (lispandfound) [reply nshmdb/fault.py:207]: Easy fix, will update.

File: pyproject.toml
Comment (lispandfound) [reply pyproject.toml]: Now!

File: nshmdb/scripts/nshm_db_generator.py
Comment (lispandfound) [reply nshmdb/scripts/nshm_db_generator.py]: yup good catch

File: nshmdb/fault.py
Comment (lispandfound) [reply nshmdb/fault.py]: fixed

**Review by sungeunbae — DISMISSED** [review]

File: nshmdb/fault.py
Comment (claudio525) [inline nshmdb/fault.py:69]: Needs to be added to the docstring, also is there a specific order to the corners? Specifically regarding which two corners define the top-edge line?

**Review by claudio525 — DISMISSED** [review]

**Review by AndrewRidden-Harper — DISMISSED** [review]

File: nshmdb/fault.py
Comment (lispandfound) [reply nshmdb/fault.py:69]: Will update, and yes there is, I'll document that in both the `corners` property and the `corners_nztm` variable.

**Review by joelridden — DISMISSED** [review]

**Comment by lispandfound** [comment]
we can probably merge this after @claudio525 and @sungeunbae approve.

**Review by claudio525 — DISMISSED** [review]

**Review by sungeunbae — DISMISSED** [review]

**Comment by lispandfound** [comment]
Sorry @claudio525, just had to switch to optional so it works on nesi :face_with_diagonal_mouth:

**Review by claudio525 — APPROVED** [review]

**Review by joelridden — APPROVED** [review]

**Pull request merged**
